apple/keyboardObserver/REAKeyboardEventObserver.mm

```diff
@@ -78,31 +82,108 @@ - (void)unsubscribeFromKeyboardEvents:(int)listenerId
 
 #else
 
+- (void)runListeners:(float)keyboardHeight
+{
+  for (NSString *key in _listeners.allKeys) {
+    ((KeyboardEventListenerBlock)_listeners[key])(_state, keyboardHeight);
+  }
+}
+
 - (void)runUpdater
 {
   [[self getDisplayLink] setPaused:NO];
-  [self updateKeyboardFrame];
+  _animtionStartTimestamp = 0;
 }
 
-- (void)updateKeyboardFrame
+- (float)getTargetTimestamp
+{
+  float targetTimestamp = _displayLink.targetTimestamp;
+  return reanimated::calculateTimestampWithSlowAnimations(targetTimestamp) * 1000;
+}
+
+- (float)estimateProgressForDuration:(float)keyboardAnimationDuration
+                                  a1:(float)a1
+                                  a2:(float)a2
+                                  b1:(float)b1
+                                  b2:(float)b2
+                                  c1:(float)c1
+                                  c2:(float)c2
 {
-  BOOL isAnimatingKeyboardChange = _measuringView.layer.presentationLayer.animationKeys.count != 0;
-  CGRect measuringFrame =
-      isAnimatingKeyboardChange ? _measuringView.layer.presentationLayer.frame : _measuringView.frame;
+  CFTimeInterval elapsedTime = _displayLink.targetTimestamp - _animtionStartTimestamp;
+  float timeProgress = elapsedTime / keyboardAnimationDuration;
+  timeProgress = fmax(fmin(timeProgress, 1), 0);
+  float x = timeProgress;
+  float progress = 1 - a1 * pow(1 - x, a2) - b1 * x * pow(1 - x, b2) - c1 * pow(x, 2) * pow(1 - x, c2);
+  return progress;
+}
+
+- (CGFloat)estimateCurrentKeyboardHeightDuringAppearing
+{
+  // Values comes from estimation: https://www.desmos.com/calculator/clhzejf5bs
+  float progress = [self estimateProgressForDuration:0.5 a1:1 a2:5.1 b1:1.6 b2:7.6 c1:0.2 c2:2.4];
+  float currentKeyboardHeight = _targetKeyboardHeight * progress;
+  return currentKeyboardHeight;
+}
+
+- (CGFloat)estimateCurrentKeyboardHeightDuringDisappearing
+{
+  // Values comes from estimation: https://www.desmos.com/calculator/d3v550ofzs
+  float progress = [self estimateProgressForDuration:0.45 a1:1 a2:5.5 b1:2.5 b2:6.4 c1:1.6 c2:3.3];
```

Comment:
Let's extract those values to some constants so we don't have to use them mutliple times like this.